measures/src/Data/Measure/ByteSize.hs

```diff
@@ -0,0 +1,328 @@
+
+-- | The types 'ByteSize8', 'ByteSize16', 'ByteSize32', and 'ByteSize64' safely
+-- capture the result of 'ByteSize' calculations that did not overflow
+-- 'ByteSize' /and/ fit in the type.
+--
+-- They intentionally have no operators! All calculations should be done in
+-- 'ByteSize'.
+newtype ByteSize8 = ByteSize8 { unByteSize8 :: Word8 }
+    -- ^ See the type's Haddock.
+  deriving stock   (Generic)
+  deriving stock   (Eq, Ord)
+  deriving newtype (Bounded)
+  deriving newtype (NFData)
+  deriving newtype (FromCBOR, ToCBOR)
+  deriving         (Read, Show)
+           via         Quiet ByteSize8
+  deriving         (NoThunks)
+           via         OnlyCheckWhnfNamed "ByteSize8" ByteSize8
+
+-- | See the documentation on 'ByteSize8'.
+newtype ByteSize16 = ByteSize16 { unByteSize16 :: Word16 }
+    -- ^ See the documentation on 'ByteSize8'.
+  deriving stock   (Generic)
+  deriving stock   (Eq, Ord)
+  deriving newtype (Bounded)
+  deriving newtype (NFData)
+  deriving newtype (FromCBOR, ToCBOR)
+  deriving         (Read, Show)
+           via         Quiet ByteSize16
+  deriving         (NoThunks)
+           via         OnlyCheckWhnfNamed "ByteSize16" ByteSize16
+
+-- | See the documentation on 'ByteSize8'
+newtype ByteSize32 = ByteSize32 { unByteSize32 :: Word32 }
+    -- ^ See the documentation on 'ByteSize8'.
+  deriving stock   (Generic)
+  deriving stock   (Eq, Ord)
+  deriving newtype (Bounded)
+  deriving newtype (NFData)
+  deriving newtype (FromCBOR, ToCBOR)
+  deriving         (Read, Show)
+           via         Quiet ByteSize32
+  deriving         (NoThunks)
+           via         OnlyCheckWhnfNamed "ByteSize32" ByteSize32
+
+-- | See the documentation on 'ByteSize8'
+newtype ByteSize64 = ByteSize64 { unByteSize64 :: Word64 }
+    -- ^ See the documentation on 'ByteSize8'.
+  deriving stock   (Generic)
+  deriving stock   (Eq, Ord)
+  deriving newtype (Bounded)
+  deriving newtype (NFData)
+  deriving newtype (FromCBOR, ToCBOR)
+  deriving         (Read, Show)
+           via         Quiet ByteSize64
+  deriving         (NoThunks)
+           via         OnlyCheckWhnfNamed "ByteSize64" ByteSize64
+
+--------------------------------------------------------------------------------
+-- Conversions
+--------------------------------------------------------------------------------
+
+-- | Types where every value can be decidably classified as one of the following.
+--
+--   - An integer within the INVARIANT interval of 'ByteSize'.
+--   - An integer greater than the INVARIANT interval of 'ByteSize'.
+--   - Neither of those. For example, negative numbers, fractions, orange, etc.
+--
+-- Law: 'partialFrom' returns 'SNothing' only in the third case above.
+--
+-- Law: 'partialFrom' returns 'SJust' an overflow if and only if the second
+-- case above.
+class ByteSizePartialFrom a where
+    -- | See 'ByteSizePartialFrom'.
+    partialFrom :: a -> StrictMaybe ByteSize
```

Comment:
I wouldn't really call it partial, since partial functions in Haskell usually are the ones that produce exceptions, eg. `error` is a partial function. I get it that math has a different view on the subject, but in order to stick to Haskell nomenclature I'd call it something like `toByteSizeMaybe`

`DefaultSignatures` can reduce some boiler plate
```suggestion
    partialFrom :: a -> StrictMaybe ByteSize
    default partialFrom :: ByteSizeFrom a => a -> StrictMaybe ByteSize
    partialFrom = partialFromDefault
```